### PR TITLE
Feature: Tokenomics TWAP Impementation

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

Include relevant motivation and context. A link to the issue is enough if it contains all the relevant 
information.

## Changes proposed in this pull request

Include a summary of the change.

## Test plan

Please describe the tests cases that you ran to verify your changes. Add further instructions on 
how to run them if needed (i.e. migration / deployment scripts, env vars, etc).

## Checklist

- [ ] Rebased the base branch
- [ ] Attached corresponding Github issue
- [ ] Prefixed the name with the type of change (i.e. feat, chore, test)
- [ ] Performed a self-review of my own code
- [ ] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [ ] Didn't generate new warnings
- [ ] Didn't generate failures on existing tests
- [ ] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/interfaces/IMasterAwareV2.sol [hunk 1]
```diff
@@ -18,11 +18,7 @@ interface IMasterAwareV2 {
     AS, // Assessment.sol
     CI, // IndividualClaims.sol - Claims for Individuals
     CG, // YieldTokenIncidents.sol - Claims for Groups
-    // TODO: 1) if you update this enum, update lib/constants.js as well
```
Comment (roxdanila) [inline contracts/interfaces/IMasterAwareV2.sol:21]: shouldn't we keep the first TODO?

File: contracts/interfaces/IRamm.sol [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  function getSpotPrices() external view returns (uint spotPriceA, uint spotPriceB);
+
+  function getBookValue() external view returns (uint bookValue);
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+  function swap(uint nxmIn)  external payable;
+
+  function addBudget(uint amount) external;
```
Comment (roxdanila) [inline contracts/interfaces/IRamm.sol:67]: remove this functionality

File: contracts/interfaces/IRamm.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+import "./IPool.sol";
+import "./INXMToken.sol";
+import "./ITokenController.sol";
+
+
+enum UintParams {
+  targetLiquidity,
+  twapDuration,
+  aggressiveLiqSpeed,
```
Comment (roxdanila) [inline contracts/interfaces/IRamm.sol:13]: suggest replacing `aggressive` with `fast`

File: contracts/interfaces/IRamm.sol [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+ struct CumulativePriceCalculationProps {
+  uint previousEthReserve;
+  uint currentEthReserve;
+  uint previousNxmA;
+  uint currentNxmA;
+  uint previousNxmB;
+  uint currentNxmB;
+  uint ratchetSpeedA;
+  uint ratchetSpeedB;
```
Comment (roxdanila) [inline contracts/interfaces/IRamm.sol:37]: is this always the current ratchet speed?

File: contracts/modules/capital/Ramm.sol [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+  uint public constant PRICE_BUFFER = 100;
+  uint public constant PRICE_BUFFER_DENOMINATOR = 10_000;
+  uint public constant WINDOW_SIZE = 172_800; // 2 days
+  uint public constant GRANULARITY = 2;
+  uint public constant PERIOD_SIZE = 86_400; // day
+
+
+  /* =========== IMMUTABLES ========== */
+
+  uint public immutable aggressiveLiquiditySpeed;
```
Comment (roxdanila) [inline contracts/modules/capital/Ramm.sol:38]: same here: aggressive -> fast

File: contracts/modules/capital/Ramm.sol [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+    a.nxmReserve = uint96(_liquidity * 1 ether / spotPriceA);
+    a.liquiditySpeed  = uint16(liqSpeedOut);
+    a.ratchetSpeed  = uint16(ratchetSpeedA);
+
+    b.nxmReserve = uint96(_liquidity * 1 ether / spotPriceB);
+    b.liquiditySpeed  = uint16(liqSpeedIn);
+    b.ratchetSpeed  = uint16(ratchetSpeedB);
+  }
+
+  function swap(uint nxmIn) external payable {
```
Comment (roxdanila) [inline contracts/modules/capital/Ramm.sol:73]: you should add a `minOut` input, to control the slippage

File: contracts/modules/capital/Ramm.sol [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+    uint _capacity = pool().getPoolValueInEth();
+
+    // TODO add buffer into calculation
+    if (_capacity < mcrETH) {
+      revert("NO_SWAPS_IN_BUFFER_ZONE");
+    }
+
+  }
+
+  function addBudget(uint amount) external override onlyGovernance {
```
Comment (roxdanila) [inline contracts/modules/capital/Ramm.sol:133]: please remove

File: contracts/interfaces/IMasterAwareV2.sol (hunk 1 above)
Comment (MilGard91) [reply contracts/interfaces/IMasterAwareV2.sol:21]: Yes, I brought it back

File: contracts/interfaces/IRamm.sol (hunk 2 above)
Comment (MilGard91) [reply contracts/interfaces/IRamm.sol:67]: Done

File: contracts/interfaces/IRamm.sol (hunk 4 above)
Comment (MilGard91) [reply contracts/interfaces/IRamm.sol:37]: For now yes, but as Rei mentioned yesterday we will add festRatchetSpeed for below pool

**Pull request merged**